### PR TITLE
Adding a4a support to TripleLift amp-ads

### PR DESCRIPTION
**Pull request by szach**
Issue #6883

### CODE REVIEW

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AmpA4A} from '../../amp-a4a/0.1/amp-a4a';
+import {base64UrlDecodeToBytes} from '../../../src/utils/base64';
+
+const AMP_SIGNATURE_HEADER = 'X-AmpAdSignature';
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: Add /** @type {string} */ here and below

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ * through a source URL and extracts the generated signature
+ * from a HTTP header.  This is then given to A4A to validate against
+ * the cloudflare signing key.
+ */
+export class AmpAdNetworkTripleliftImpl extends AmpA4A {
+
+  /**
+   * @param {!Element} element
+   */
+  constructor(element) {
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: Given you're only calling super, why is this necessary?

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  /** @override */
+  getAdUrl() {
+    return this.element.getAttribute('src').replace(TRIPLELIFT_BASE_URL, TRIPLELIFT_BASE_A4A_URL);
+  }
+
+  /** @override */
+  extractCreativeAndSignature(responseText, responseHeaders) {
+    var signature = null;
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: replace var with let

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  /**
+   * @param {!Element} element
+   */
+  constructor(element) {
+    super(element);
+  }
+
+  /** @override */
+  isValidElement() {
+    return this.element.hasAttribute('src') && this.isAmpAdElement();
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: you should check for the src attribute within isA4A predicate as isValidElement returning false will cause the ad request to be dropped meaning it does not fallback to 3p

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  /** @override */
+  getSigningServiceNames() {
+    return ['cloudflare'];
+  }
+
+  /** @override */
+  getAdUrl() {
+    return this.element.getAttribute('src').replace(TRIPLELIFT_BASE_URL, TRIPLELIFT_BASE_A4A_URL);
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: Do you want to verify that the src attribute has the TRIPLELIFT_BASE_URL to start? I noticed that you're 3p implementation includes validateSrcPrefix('https://ib.3lift.com/'.  Without it, this could allow a callout to ANY XHR CORS end point (e.g. src to doubleclick).  This is supplied by the publisher so what if it includes javascript(sendMeYourCookies())?

File: extensions/amp-ad-network-triplelift-impl/0.1/triplelift-a4a-config.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+/**
+ * @param {!Window} win
+ * @param {!Element} element
+ * @returns {boolean}
+ */
+export function tripleliftIsA4AEnabled(win, element) {
+  const a4aRequested = element.getAttribute('data-use-a4a');
+  return !!a4aRequested;
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-triplelift-impl/0.1/triplelift-a4a-config.js]: Why not just 
`return !!element.getAttribute('data-use-a4a');`

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js (hunk 2 above)
Comment (oliy) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: This was actually in our template code as a placeholder, to show where initialization code could go.  I'm adding comments to the template code that this is an optional block and can be removed if there is no content.

**Comment by szach** [comment]
Thanks @keithwrightbos @oliy - this was admittedly rushed to try and make yesterday's canary. Will look to address all these points and implement a fallback to our non-a4a implementation in advance of next week's cut.

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js (hunk 1 above)
Comment (szach) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: 👌

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js (hunk 2 above)
Comment (szach) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: 👌

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js (hunk 4 above)
Comment (szach) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: @keithwrightbos - where exactly are you proposing this should be implemented if at all?

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js (hunk 5 above)
Comment (szach) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: 👌

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js (hunk 3 above)
Comment (szach) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: 👌

File: extensions/amp-ad-network-triplelift-impl/0.1/triplelift-a4a-config.js (hunk 6 above)
Comment (szach) [inline extensions/amp-ad-network-triplelift-impl/0.1/triplelift-a4a-config.js]: 👌

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js (hunk 4 above)
Comment (keithwrightbos) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: I would suggest at least the same check as you have in your current 3p implementation (https://github.com/ampproject/amphtml/blob/master/ads/triplelift.js):
`validateSrcPrefix('https://ib.3lift.com/', src);`

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js (hunk 5 above)
Comment (keithwrightbos) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: I would add check in the config use A4A predicate as well as isValidElement.  Both will ensure proper value well before getAdUrl would execute.

**Comment by keithwrightbos** [comment]
I have reopened the PR.  Sorry I did not see notifications because the PR was closed.
@lannka & @jridgewell to look ASAP given we are hoping to get into today's canary cut

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js (hunk 4 above)
Comment (szach) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: 👌

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js (hunk 5 above)
Comment (szach) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: 👌

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+ * This is a minimalistic AmpA4A implementation that primarily gets an Ad
+ * through a source URL and extracts the generated signature
+ * from a HTTP header.  This is then given to A4A to validate against
+ * the cloudflare signing key.
+ */
+export class AmpAdNetworkTripleliftImpl extends AmpA4A {
+
+  /** @override */
+  isValidElement() {
+    return this.isAmpAdElement();
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: Let's add check for src attribute and starting with base url you expect just to be safe.

File: extensions/amp-ad-network-triplelift-impl/0.1/test/test-amp-ad-network-triplelift-impl.js [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+          return ['cloudflare'];
+        });
+    tripleliftImpl = new AmpAdNetworkTripleliftImpl(tripleliftImplElem);
+  });
+
+  afterEach(() => {
+    sandbox.restore();
+  });
+
+  describe('#isValidElement', () => {
```
Comment (keithwrightbos) [inline extensions/amp-ad-network-triplelift-impl/0.1/test/test-amp-ad-network-triplelift-impl.js]: Can we add a negative test here (create instance of amp-ad-network-triplelift element instead of amp-ad) as well as checks regarding src.

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js (hunk 7 above)
Comment (szach) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: since we’re checking for src attribute and base url in tripleliftIsA4AEnabled is there a reason to check it again here?

File: extensions/amp-ad-network-triplelift-impl/0.1/test/test-amp-ad-network-triplelift-impl.js (hunk 8 above)
Comment (szach) [inline extensions/amp-ad-network-triplelift-impl/0.1/test/test-amp-ad-network-triplelift-impl.js]: added the negative test; but again,, since we’re checking for src attribute and base url in tripleliftIsA4AEnabled is there a reason to check it again here?

File: extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js (hunk 7 above)
Comment (keithwrightbos) [inline extensions/amp-ad-network-triplelift-impl/0.1/amp-ad-network-triplelift-impl.js]: I was being overly cautious... we do not need to include it

File: extensions/amp-ad-network-triplelift-impl/0.1/test/test-amp-ad-network-triplelift-impl.js (hunk 8 above)
Comment (keithwrightbos) [inline extensions/amp-ad-network-triplelift-impl/0.1/test/test-amp-ad-network-triplelift-impl.js]: Agreed

File: extensions/amp-ad-network-triplelift-impl/0.1/test/test-amp-ad-network-triplelift-impl.js [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+  AmpAdXOriginIframeHandler, // eslint-disable-line no-unused-vars
+} from '../../../amp-ad/0.1/amp-ad-xorigin-iframe-handler';
+import {base64UrlDecodeToBytes} from '../../../../src/utils/base64';
+import {utf8Encode} from '../../../../src/utils/bytes';
+import * as sinon from 'sinon';
+import {tripleliftIsA4AEnabled} from '../triplelift-a4a-config';
+import {createElementWithAttributes} from '../../../../src/dom';
+import {createIframePromise} from '../../../../testing/iframe';
+
+describe.only('triplelift-a4a-config', () => {
```
Comment (lannka) [inline extensions/amp-ad-network-triplelift-impl/0.1/test/test-amp-ad-network-triplelift-impl.js]: remove `only`
